src/framework/core/js/ModelTransformationTransforms.js

```diff
@@ -349,86 +349,182 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
-            fluid.fail("setMembershipToArray requires an options block specified");
-        }
-
-        if (transformSpec.presentValue === undefined) {
-            transformSpec.presentValue = true;
-        }
-
-        if (transformSpec.missingValue === undefined) {
-            transformSpec.missingValue = false;
+    /**
+     *
+     * Transform the input object into an array based upon the options provided to the specification. Accepts
+     * an object as its first argument and an optional second argument (an object) which contains the
+     * specifications for the resulting output.
+     *
+     * An error will be thrown if the input to be transformed is not an object.
+     *
+     * Partly invertible (into arrayToSetMembership)
+     *
+     * For example, with the following transformSpec:
+     *
+     * {
+     *   "options": {
+     *      "hasMouse": "mouse",
+     *      "hasKeyboard": "keyboard",
+     *      "hasTrackpad": "trackpad",
+     *      "hasHeadtracker": "headtracker"
+     *   },
+     *   "presentValue": "supported",
+     *   "missingValue": "not supported"
+     * }
+     *
+     * For an input object like this:
+     *
+     * {
+     *   "hasMouse": "supported",
+     *   "hasKeyboard": "supported",
+     *   "hasTrackpad": "not supported",
+     *   "hasHeadtracker": "not supported"
+     * }
+     *
+     * The output array will be:
+     *
+     * ["keyboard", "mouse"]
+     *
+     * https://docs.fluidproject.org/infusion/development/ModelTransformationAPI.html#fluidtransformssetmembershiptoarray
+     *
+     * @param {Object} input - The object to be transformed into an array.
+     * @param {Object} [transformSpec] - The options provided to the transformation rule.
+     * @return {Array} - The transformed array
+     *
+     */
+    fluid.transforms.setMembershipToArray = function (input, transformSpec) {
+        // <input> should be an object.
+        if (!input || !fluid.isPlainObject(input, true)) {
+            fluid.fail("setMembershipToArray didn't find object at inputPath nor passed as value.");
         }
-
         var outputArr = [];
+        var presentValue = fluid.get(transformSpec, "presentValue") ? transformSpec.presentValue : true,
+            options = fluid.get(transformSpec, "options");
+
+        // When <options> are provided in the transformation rule, use them for entries in the output array.
         fluid.each(options, function (outputVal, key) {
-            var value = fluid.get(input, key, transformer.resolverGetConfig);
-            if (value === transformSpec.presentValue) {
+            // Write to output array the key of the <input> object if it's value matches the <presentValue>
+            if (fluid.get(input, key) === presentValue) {
                 outputArr.push(outputVal);
             }
         });
+
+        // Otherwise, when no <options> are provided, use the <input> object for entries in the output array.
+        if (!options) {
```

Comment:
As above, I think a full if () {} else {} structure is more logically clear

Reply:
Hi @amb26, thanks for pointing this out. Although the current code is logically correct, I agree that it would make more sense to use an if-else statement and be more apparent to someone trying to read it.